modules/dnn/src/opencl/softmax_loss.cl

```diff
@@ -0,0 +1,303 @@
+  for (int index = get_global_id(0), s = 0; index < spatial_dim * get_local_size(0); index +=
+      get_global_size(0), ++s) {
+    float maxval = -FLT_MAX;
+    for (int c = get_global_id(0); c < channels; c += get_global_size(0)) {
+      Dtype tmp = data[(n * channels + c) * spatial_dim + s];
+      maxval = max((Dtype)tmp, (Dtype)maxval);
+    }
+    maxval = sub_group_reduce_max(maxval);
+    //if (get_sub_group_local_id() == 0)
+    group_tmp[get_sub_group_id() * spatial_dim + s] = maxval;
+  }
+  barrier(CLK_GLOBAL_MEM_FENCE);
+
+  for (int index = get_global_id(0); index < spatial_dim * get_max_sub_group_size(); index +=
+      get_global_size(0)) {
+    int s = index / get_max_sub_group_size();
+    Dtype maxval = sub_group_reduce_max(group_tmp[get_sub_group_local_id() * spatial_dim + s]);
+    //if (get_sub_group_local_id() == 0)
+    scale[n * spatial_dim + s] = maxval;
+  }
+
+  barrier(CLK_GLOBAL_MEM_FENCE);
+
+  for (int index = get_global_id(0); index < channels * spatial_dim;
+      index += get_global_size(0)) {
+    int s = index % spatial_dim;
+    out[n * channels * spatial_dim + index] = exp(data[n * channels * spatial_dim + index] - scale[n * spatial_dim + s]);
+  }
+  barrier(CLK_GLOBAL_MEM_FENCE);
+
+  for (int index = get_global_id(0), s = 0; index < spatial_dim * get_local_size(0); index +=
+      get_global_size(0), ++s) {
+    Dtype sum = 0;
+    for (int c = get_global_id(0); c < channels; c += get_global_size(0)) {
+      sum += out[n * channels * spatial_dim + c * spatial_dim + s];
+    }
+    sum = sub_group_reduce_add(sum);
+    group_tmp[get_sub_group_id() * spatial_dim + s] = sum;
+  }
+  barrier(CLK_GLOBAL_MEM_FENCE);
+
+  for (int index = get_global_id(0); index < spatial_dim * get_max_sub_group_size(); index +=
+      get_global_size(0)) {
+    int s = index / get_max_sub_group_size();
+    Dtype sum = sub_group_reduce_add(group_tmp[get_sub_group_local_id() * spatial_dim + s]);
+    //if (get_sub_group_local_id() == 0)
+    scale[n * spatial_dim + s] = sum;
+  }
+  barrier(CLK_GLOBAL_MEM_FENCE);
+
+  for (int index = get_global_id(0); index < channels * spatial_dim;
+      index += get_global_size(0)) {
+    int s = index % spatial_dim;
+    out[n * channels * spatial_dim + index] /= scale[n * spatial_dim + s];
+  }
+}
+
+// Copied from caffe.pb.h, must keep consistent with the original definition
+#ifndef __SOFTMAX_LOSS_CL__
+#define __SOFTMAX_LOSS_CL__
+enum LossParameter_NormalizationMode {
+  LossParameter_NormalizationMode_FULL = 0,
+  LossParameter_NormalizationMode_VALID = 1,
+  LossParameter_NormalizationMode_BATCH_SIZE = 2,
+  LossParameter_NormalizationMode_NONE = 3
+};
+#endif
+// Copied from softmax_loss_layer.cpp, must keep consistent with the original implementation
+Dtype TEMPLATE(get_normalizer, Dtype)(
+    enum LossParameter_NormalizationMode normalization_mode, int valid_count,
+    int outer_num_, int inner_num_) {
+  Dtype normalizer;
+  switch (normalization_mode) {
```

Comment:
`normalization_mode` should be passed via build macro too

Reply:
fixed in new updates.